notes/radicle-p2p-hypercore-overview.md

```diff
@@ -0,0 +1,114 @@
+
+[hypercore]: https://github.com/mafintosh/hypercore
+[discovery-swarm]: https://github.com/mafintosh/discovery-swarm
+
+
+Hypercore
+---------
+
+`hypercore` is a library to distribute cryptographically secure append-only logs
+called *feeds*. A feed consists of a list of *entries* which may be JSON objects
+or binary blobs. A feed is signed with a private key. The hash of the
+corresponding public key is called the *discovery key* and acts as an identifier
+for the feed.
+
+Any user who knows the location (for example the IP and port) of a feed can read
+any entry in the feed, verify the integrity of any entry, listen to new entries,
+and replicate the feed.
+
+We say a user *hosts a feed* if the user makes it available to the world. If a
+users hosts a feed the user is either the owner of that feed or she replicates
+it.
+
+`hypercore` supports feed subscriptions. A user can subscribe to a feed and will
+be notified if the owner publishes a new entry.
+
+Every feed a user owns or replicates is stored locally and updated automatically
+if a user goes online.
+
+`discovery-swarm` is a library to discover hosts for hypercore feeds and
+announce the hosting of a feed. It uses the Bittorrent DHT and centralised DNS
+for global discovery as well as Multicast DNS for local discovery.
+
+
+Single-writer Chains
+--------------------
+
+*Single-writer chains* are Radicle chains that only the owner of a specific
+private key can write to.
+
+We use hypercore feeds to represent single-writer chains. Each entry in the feed
+is a Radicle expression. Single-writer chains are identified by the discovery
+key of their feed.
+
+To read a Radicle chain a user needs the discovery key of a feed. With the
+discovery key the user uses hyperdiscovery to find someone who replicates the
+feed. (This may be the owner.) The user then reads all entries in the feed with
+hypercore and evaluates them with radicle.
+
+To write an expression to a Radicle chain a user publishes a new entry to the
+`hypercore` feed.
+
+
+Multi-writer Chains
+-------------------
+
+*Multi-writer chains* are Radicle chains that the owners of multiple distinct
+private keys can write to.
+
+We implement a multi-writer chain by using set of feeds owned by different
+users. Every entry in these feeds is a Radicle expression and a link to another
+entry in any of feeds. Every entry in a feed now determines a linked list of
+Radicle expression (assuming the feed entries and links are well formed and
+valid) and thus a Radicle chain.
+
+To derive a Radicle chain from a set of feeds we need an algorithm that takes
+the heads of all the feeds of a multi-writer chain and outputs either a chain of
+Radicle expressions or signals that the chain is invalid.
+
+The simplest such algorithm is the following: We accept only the situation where
+there is a feed with head entry `e` so that the head entries of all other feeds
+are ancestors of `e`. In that case `e` is the chain derived from the set of
```

Comment:
```suggestion
are either `e` or ancestors of `e`. In that case `e` is the tip chain derived from the set of
```

Reply:
> are either `e` or ancestors of `e`.

I don’t think there is a sensible notion of equality of entries in different feeds that would support this statement. Entries are addressed by their feed key and the sequence number. So by definition entries in different feeds will have different addresses. One could say that entries are the same if there content is the same. But this would require every collaborator to replay entries from every other collaborator to their chain.

> the tip chain

What does the term “tip chain” mean?

Reply:
I see your point about ancestors. I was thinking we were just porting over the IPFS idea, where it is possible for the multiple IPNSs to point to exactly the same hash, but I like your idea better.

> What does the term “tip chain” mean?

 I think `head` might be the term you're using. But as it stands, it seems wrong to me: `e` isn't the chain itself, it's just the latest entry in it.

Reply:
> > What does the term “tip chain” mean?
> 
> I think `head` might be the term you're using. But as it stands, it seems wrong to me: `e` isn't the chain itself, it's just the latest entry in it.

Yeah, right. `e` refers to an entry. I’ll reformulate it.